### ISSUE TITLE
RFC: Replace transport's mutex with message passing

### ISSUE THREAD

**Issue by kureuil** [sozu-proxy/lapin#148]
> Disclaimer: I fully expect to do the work for this and don't want to force anyone to take on this work. I'm posting this here to make sure this change is desired and to refine its final form.

## Summary

Put the `AMQPTransport` in a background task, and have `Client` and `Channel` communicate with it using `Future`-aware message channels.

## Motivation

Lapin is a collection of libraries to communicate with RabbitMQ servers, one of which provides asynchronous networking via the `Future` abstraction popular in the Rust ecosystem. Because a `Future` will often be sent across threads, it is important to be able to fulfill this contract while still maintaining good ergonomics for the library. To this end, the transport (an abstract above network sockets dealing with frames) is wrapped in an `Arc<Mutex<T>>`, making it shareable & mutable across threads. However this architecture creates two main problems:

1. Future resource handling is at odds with Mutex semantics. The goal of a Mutex is to share a resource between multiple threads, whereas `Future`-aware resources are meant to be owned by a task. This means that when multiple tasks register interest for a resource, only the task [that registered last will be notified]. This leads to tasks deadlocking and blocking the program execution. It is possible to circumvent this problem by manually storing a vector of tasks and re-dispatch notifications to these tasks but it's tedious, error-prone work that won't solve the second problem.
2. It is impossible to implement a proper backpressure mechanism. Because a `Future`-aware mutex doesn't (can't ?) exist, it is not possible to wake tasks and guarantee that it will be able to make progress. Indeed, the standard mutex behavior means putting the thread to sleep which poses problem with scheduling w.r.t future exectutors that don't [expect this behavior] and can create unnecessary contention. Instead, it should be possible to be notified once a future is able to make progress again.

The proposed solution would be to rework the internals of the library to move the transport into a background task, and have the `Client` and `Channel` structs use message channels to communicate with it. In the background task, there would be a `mpsc::Receiver<(TBD, oneshot::Sender<TBD2>)>` listening for commands to send to the RabbitMQ server, sending the commands, and notifying the task waiting for its answer using the `oneshot::Sender` given with the request.

Let's see how this solves the two aforementioned problems:

1. *Future resource handling is at odds with Mutex semantics*: `Future`-aware message channels provided by the `futures` crate are (unsurprisingly) designed with future resource handling. On the contrary of `Mutex`, `mpsc::Sender` are meant to be owned and easily clonable (with two clones being two different resource future) simplifying a lot the scheduling model of the library.
2. *It is impossible to implement a proper backpressure mechanism*: again, the message channel implementations provided by the `futures` crate are built with backpressure in mind. If a task is parked because a `mpsc::Sender` returns `Async::NotReady` because the channel is full, it is guaranteed the it will be able to make progress when it will be notified. Also, because its implementation respects the contract established by the `Future` trait, this means that waiting for the `mpsc::Sender` to be ready won't put the thread to sleep, letting the future executor schedule other futures properly.

Refactoring the library to use message passing also brings other changes:

* It would be possible to drop the `T: AsyncWrite + AsyncRead + Sync + Send + 'static` bound on the `Channel` & `Client` structs, making them a even more ergonomic to use.
* It would be easier to test the parts in isolation: no low-level parsing logic when testing if a command is sent to the background task, and no meddling with high-level code when testing command serialization.

[that registered last will be notified]: https://docs.rs/futures/0.1.25/futures/future/trait.Future.html#return-value
[expect this behavior]: https://docs.rs/futures/0.1.25/futures/future/trait.Future.html#tymethod.poll

## Drawbacks

This change implies reworking a lot of the lapin-futures crate internals, and adds an indirection layer between the high-level structs and the network socket, but it should be considered more important to be able to schedule the tasks correctly than it is to be performant at the expense of correctness.

## Prior art

* Trust-DNS AyncResolver uses a background task to share DNS resolution work between multiple handles, making it easier to use in a multi-threaded environment.
  *Source:* https://github.com/bluejekyll/trust-dns/pull/487
* Pleingres (Asynchronous PostgresQL driver) spawns a task owning the network socket to the database and returns handles to this task to the end-users.
  *Source:* https://docs.rs/pleingres/0.4.1/pleingres/
* The tokio/doc-push repository outline of the concurrency section hints at the fact that mutexes don't work in an asynchronous context.
  *Source*: https://github.com/tokio-rs/doc-push/blob/master/outline/concurrency.md#contents-1

## Unresolved questions

What should be sent over the mpsc channel between the background task and the `Client` and `Channel` struct ? A boxed trait object (maybe even a reference once std::future + pinning lands in stable ?) would probably be the best solution due to the, sometimes pretty massive, differences in behavior between commands but I'm not completely sure it is the best way forward.

**Comment by Keruspe** [comment]
I did not read the whole thing yet, only parts of it, but it seems to really match what my short term plan was "when I get enough time".

What I planned on doing:

First, in lapin-async:
- Write a build.rs using amq-protocol codegen to generate an enum with all the amqp verbs and option structs
- Add a function which takes this enum as a parameter and dispatch to the right function

Then, in lapin-futures:
- Write a function which takes the enum and pass it to lapin-async
- Turn AMQTransport into a future which sequentially deals with "Action" we pass it (basically, the same enum from above)
- Drop the now unused mutex

I first tried to use futures-locks but the performance was terrible. I intended to do some more benchmark comparing:
- the current version
- using parking_lot as a mutex implementation
- using the scheme described above

**Comment by Keruspe** [comment]
I still haven't found the time to start working on this yet, help greatly appreciated.

**Comment by jamwaffles** [comment]
I don't think I'm experienced enough with Rust yet to take a crack at this, but I've run into an issue which I think it might solve. I've tried to use the same `Channel` to emit two messages at the "same" time using `channel.basic_publish(/* ... */).join(channel.basic_publish(/* .... */))`. I expect two messages to be sitting on the queue, but only see one. Presumably this is something the above RFC would fix. Should I open a separate issue with my problem, linking to this RFC?

**Comment by kureuil** [comment]
*Sorry for having gone silent, $life got in the way as it loves to.*

@jamwaffles I'd love to help you solve your problem, could you open an issue with a minimal code reproducing the issue & context information (OS, Rust version, RabbitMQ server version, etc.) ? The work resulting from this RFC might solve this problem but I'm surprised it is one to begin with.

@Keruspe

> First, in lapin-async:
>
> * Write a build.rs using amq-protocol codegen to generate an enum with all the amqp verbs and option structs
> * Add a function which takes this enum as a parameter and dispatch to the right function

This seems like a reasonnable change, though I'd advocate for the codegen to not be put in the `build.rs`: the generated code should always be the same so it seems safer to generate it once and commit the generated file to the repository.

> Then, in lapin-futures:
> 
> * Write a function which takes the enum and pass it to lapin-async
> * Turn AMQTransport into a future which sequentially deals with "Action" we pass it (basically, the same enum from above)
> * Drop the now unused mutex

I'm not sure how you're getting from step 2 to step 3, wouldn't `AMQPTransport` still need to borrow mutably its inner `lapin_async::Connection` instance making the `Mutex` mandatory in a multi-threaded environment ? Or maybe I'm just not understanding what the second step consists of ?

**Comment by jamwaffles** [comment]
Thanks for the response! I've been trying to reproduce my issue in a clean repo with the minimal amount of code copied over and can't replicate the missing event behaviour, so I'll hold off on opening an issue. I want to see what this RFC might do. I'm also looking at porting my code to the nightly async/await syntax which may also help.

**Comment by xabufr** [comment]
Maybe we should look at something like https://github.com/tower-rs/tokio-tower (when it will be stable) ?

**Comment by Keruspe** [comment]
That’s definitely not something we want to do. Doesn’t fit our design at all

**Comment by Keruspe** [comment]
Fwiw, I’m working on a rework/refactor that will make this a lot easier. Should land by the end of the month

**Comment by Keruspe** [comment]
0.19.0-alpha1 has some work towards this. The global mutex in futures is used a lot less. There are several smaller ones in async now though.

**Comment by Keruspe** [comment]
AMQPTransport is now gone
